classes/Village.php

```diff
@@ -38,6 +58,34 @@ public function __construct($system, $village = '', array $village_row = []) {
             $this->coords = VillageManager::getLocation($this->system, $this->village_id);
             $this->relations = VillageManager::getRelationsForVillage($this->system, $this->village_id);
             $this->policy = new VillagePolicy($this->policy_id);
+            $this->buildings = VillageUpgradeManager::getBuildingsForVillage($this->system, $this->village_id);
+            $this->upgrades = VillageUpgradeManager::getUpgradesForVillage($this->system, $this->village_id);
+            $this->active_upgrade_effects = VillageUpgradeManager::initializeEffectsForVillage($this->system, $this->upgrades);
+            $this->construction_speed += ($this->active_upgrade_effects[VillageUpgradeConfig::UPGRADE_EFFECT_CONSTRUCTION_SPEED] / 100);
+            $this->research_speed += ($this->active_upgrade_effects[VillageUpgradeConfig::UPGRADE_EFFECT_RESEARCH_SPEED] / 100);
+            $this->resource_capacity += $this->active_upgrade_effects[VillageUpgradeConfig::UPGRADE_EFFECT_RESOURCE_CAPACITY];
+            if ($this->policy->construction_speed > 0) {
```

Comment:
This policy benefits to research/construction are based on how many villages have more total research/construction.